default/python/AI/Trait/Character.py

```diff
@@ -0,0 +1,687 @@
+        research_index = self.id % 2
+        if self.aggression >= fo.aggression.aggressive:
+            research_index = 2 + (self.id % 3)  # so indices [2,3,4]
+        elif self.aggression >= fo.aggression.typical:
+            research_index += 1
+        return research_index
+
+
+class Character(Trait):
+    """
+    A collection of behaviours.
+
+    For each query that Trait supports a Character queries
+    all of its own traits to determine if an action is permissible or prefered.
+    """
+
+    def __init__(self, traits):
+        self.traits = traits
+        if not all([isinstance(x, Trait) for x in traits]):
+            raise TypeError("All traits must be sub-classes of Trait")
+
+    def get_trait(self, type_of_trait):
+        """Return the requested trait or None"""
+        trait = [x for x in self.traits if isinstance(x, type_of_trait)]
+        return trait[0] if trait else Trait()
+
+
+# Complete the Character class by adding all of the combiners to combine the outputs of the
+# individual traits.  Character tries to combine results in the way most limiting to the AI
+
+def _make_single_function_combiner(funcnamei, f_combo):
+    """Make a combiner that collects the results of funcname from each trait
+    and applies f_combo to the results"""
+    def func(self, *args, **kwargs):
+        """Apply funcnamei to each trait and combine them with ''f_combo''"""
+        return f_combo([getattr(x, funcnamei)(*args, **kwargs) for x in self.traits])
+    return func
+
+# Create combiners for traits that all must be true
+for funcname in ["may_explore_system", "may_surge_industry", "may_maximize_research", "may_invade",
+                 "may-invade_with_bases", "may_build_building", "may_produce_troops",
+                 "may_dither_focus_to_gain_research", "may_research_heavily",
+                 "may_travel_beyond_supply", "may_research_xeno_genetics_variances",
+                 "prefer_research_defensive", "prefer_research_low_aggression", "may_research_tech",
+                 "may_research_tech_classic"]:
+    setattr(Character, funcname, _make_single_function_combiner(funcname, all))
+
+# Create combiners for traits that take min result
+for funcname in ["max_number_colonies", "invasion_priority_scaling",
+                 "military_priority_scaling", "max_defense_portion"]:
+    setattr(Character, funcname, _make_single_function_combiner(funcname, min))
+
+# Create combiners for traits that take max result
+for funcname in ["target_number_of_orbitals", "military_safety_factor", "get_research_index"]:
+    setattr(Character, funcname, _make_single_function_combiner(funcname, max))
+
+# Create combiners for traits that take any result
+for funcname in ["check_orbital_production"]:
+    setattr(Character, funcname, _make_single_function_combiner(funcname, any))
+
+
+# Create combiners for traits that averages all not None results
+def average_not_none(llin):
+    ll = [x for x in llin if x]
+    return sum(ll) / len(ll)
```

Comment:
Check for None explicitly. The list may contain zeros.

E.g. average of (0, 1) is 0.5 but your function returns 1.